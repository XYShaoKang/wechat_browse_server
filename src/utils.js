import jwt from 'jsonwebtoken'
import { AuthenticationError } from 'apollo-server-koa'

const APP_SECRET = process.env.APP_SECRET

const getUserId = authorization => {
  const token = authorization.replace('Bearer ', '')
  const { userId } = jwt.verify(token, APP_SECRET)
  return userId
}

const authenticated = next => (root, args, context, info) => {
  if (!context.currentUser) {
    throw new AuthenticationError(`Unauthenticated!`)
  }

  return next(root, args, context, info)
}

const generateToken = obj => jwt.sign(obj, APP_SECRET)

module.exports = {
  APP_SECRET,
  getUserId,
  authenticated,
  generateToken,
}
