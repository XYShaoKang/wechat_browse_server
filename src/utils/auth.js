import jwt from 'jsonwebtoken'

/**
 * @param {string} authorization
 * @returns {{userId:string,weChatId:string}}
 */
export const getUserId = authorization => {
  const token = authorization.replace('Bearer ', '')
  /**
   * @type {object}
   */
  const result = jwt.verify(token, process.env.APP_SECRET || '')
  return result
}

/**
 * @param {string | object | Buffer} obj
 */
export const generateToken = obj => {
  return jwt.sign(obj, process.env.APP_SECRET || '')
}
