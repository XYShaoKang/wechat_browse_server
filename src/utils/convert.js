import * as R from 'ramda'

const objOfCreate = R.objOf('create')
const objOfCreateUrl = R.pipe(
  R.objOf('url'),
  objOfCreate,
)
const createWeChatUserCreateInput = R.map(
  R.converge(R.assocPath(['avatar']), [
    R.pipe(
      R.prop('avatar'),
      R.when(
        R.is(Object),
        R.pipe(
          R.map(objOfCreateUrl),
          objOfCreate,
        ),
      ),
    ),
    R.identity,
  ]),
)

const pickUserName = R.reduce(
  R.useWith(R.concat, [
    R.identity,
    R.converge(R.concat, [
      R.propOr([], 'memberList'),
      R.pipe(
        R.prop('owner'),
        R.ifElse(R.is(String), R.of, R.always([])),
      ),
    ]),
  ]),
  [],
)

/**
 * @param {import("../../generated/prisma-client").WeChatUser[]} weChatUsers
 */
const createChatRoomCreateInput = weChatUsers => {
  const objOfCreate = R.objOf('create')
  const objOfConnect = R.objOf('connect')
  const objOfCreateUrl = R.pipe(
    R.objOf('url'),
    objOfCreate,
  )
  const notEmpty = R.pipe(
    R.either(R.isNil, R.isEmpty),
    R.not,
  )
  const ifElseFlaseToNull = R.curry((condition, onTrue) =>
    R.ifElse(condition, onTrue, R.always(null)),
  )
  const emptyToNull = ifElseFlaseToNull(notEmpty)
  const convertAvatar = R.converge(R.assocPath(['avatar']), [
    R.pipe(
      R.prop('avatar'),
      emptyToNull(
        R.pipe(
          R.map(objOfCreateUrl),
          objOfCreate,
        ),
      ),
    ),
    R.identity,
  ])
  /**
   * @param {string} uname
   */
  const mapUserNameToId = uname => {
    const user = weChatUsers.find(({ username }) => username === uname)
    if (user) {
      return user.id
    }
    return null
  }

  const objOfConnectUserId = R.pipe(
    mapUserNameToId,
    emptyToNull(R.pipe(R.objOf('id'))),
  )

  const convertOwner = R.converge(R.assocPath(['owner']), [
    R.pipe(
      R.prop('owner'),
      objOfConnectUserId,
      emptyToNull(objOfConnect),
    ),
    R.identity,
  ])

  const covertMemberList = R.converge(R.assocPath(['memberList']), [
    R.pipe(
      R.prop('memberList'),
      emptyToNull(
        R.pipe(
          R.map(objOfConnectUserId),
          emptyToNull(objOfConnect),
        ),
      ),
    ),
    R.identity,
  ])

  const convertUserName = R.pipe(
    convertOwner,
    covertMemberList,
  )
  return R.map(
    R.pipe(
      convertAvatar,
      convertUserName,
    ),
  )
}

const createMessageInput = R.map(
  R.pipe(
    R.converge(R.assocPath(['talker']), [
      R.pipe(
        R.prop('talker'),
        R.ifElse(
          R.both(R.is(String), R.complement(R.test(/@chatroom$/))),
          R.pipe(
            R.objOf('username'),
            R.objOf('connect'),
          ),
          R.always(null),
        ),
      ),
      R.identity,
    ]),
    R.converge(R.assocPath(['chatRoom']), [
      R.pipe(
        R.propOr(null, 'chatRoom'),
        R.when(
          R.is(String),
          R.pipe(
            R.objOf('username'),
            R.objOf('connect'),
          ),
        ),
      ),
      R.identity,
    ]),
  ),
)
export {
  createWeChatUserCreateInput,
  pickUserName,
  createChatRoomCreateInput,
  createMessageInput,
}
