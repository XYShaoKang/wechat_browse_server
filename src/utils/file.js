import path from 'path'
import fs from 'fs'
import crypto from 'crypto'
import mime from 'mime-types'
import { prisma } from '../../generated/prisma-client'

const STATIC_PATH = path.join(__dirname, '../../public/static/')

if (!fs.existsSync(STATIC_PATH)) {
  fs.mkdirSync(STATIC_PATH, { recursive: true })
}

const randomMd5 = () => {
  const hash = crypto.createHash('md5')
  hash.update('' + new Date().getTime() + Math.random() * 10000)
  return hash.digest('hex')
}

/**
 * @param {string} str
 */
const strMd5 = str => {
  const hash = crypto.createHash('md5')
  hash.update(str)
  return hash.digest('hex')
}

/**
 * @param {Promise<{ stream: NodeJS.ReadableStream; mimetype: string; }> | { stream: NodeJS.ReadableStream; mimetype: string; } } file
 * @returns {Promise<{ fileName: string; mimetype: string; size: number;}>}
 */
const saveFile = async file => {
  const { stream, mimetype } = await file
  const ext = mime.extension(mimetype)
  const hash = crypto.createHash('md5')
  const tempName = `${randomMd5()}.${ext}`
  const ws = fs.createWriteStream(path.join(STATIC_PATH, tempName))

  stream.on('data', chunk => {
    hash.update(chunk)
  })

  stream.pipe(ws)
  let size = 0
  stream.on('data', chunk => {
    size += chunk.length
  })

  return new Promise(function(resolve, reject) {
    ws.on('finish', err => {
      if (err) reject(err)

      const md5Str = hash.digest('hex')
      const fileName = `${md5Str}.${ext}`

      if (fs.existsSync(path.join(STATIC_PATH, fileName))) {
        fs.unlinkSync(path.join(STATIC_PATH, tempName))
        resolve({ fileName, mimetype, size })
      } else {
        fs.rename(
          path.join(STATIC_PATH, tempName),
          path.join(STATIC_PATH, fileName),
          err => {
            if (err) {
              // eslint-disable-next-line no-console
              console.log(err)
              reject(err)
            }
            resolve({ fileName, mimetype, size })
          },
        )
      }
    })
  })
}
/**
 *
 * @param {Object} fileIndex
 * @param {string} fileIndex.fileName
 * @param {string} fileIndex.mimetype
 * @param {number} fileIndex.size
 * @param {string} fileIndex.url
 * @returns {Promise<string>} id
 */
const saveToFileIndex = async ({ fileName, mimetype, size, url }) => {
  const [file] = await prisma.fileIndexes({ where: { fileName } })

  if (file) {
    return file.id
  }

  const { id } = await prisma
    .createFileIndex({ fileName, mimetype, size, url })
    .catch(err => {
      throw new Error(err)
    })
  return id
}

export { STATIC_PATH, strMd5, saveFile, randomMd5, saveToFileIndex }
