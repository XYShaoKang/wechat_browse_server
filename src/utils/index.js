export * from './auth'
export * from './convert'
