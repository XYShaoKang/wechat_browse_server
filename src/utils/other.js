import * as R from 'ramda'
import { Observable, from, of } from 'rxjs'
import { mergeAll, map, mergeMap } from 'rxjs/operators'

/**
 * @param {(key: string,parent:any) => any} asyncFn
 * @param {Object} data
 * @param {[import('rxjs').Observable<any>]} observables
 */
export function _asyncMap(asyncFn, data, observables) {
  for (const key in data) {
    if (data.hasOwnProperty(key)) {
      const result = asyncFn(key, data)
      if (result instanceof Observable) {
        observables.push(result)
      } else if (result instanceof Promise) {
        observables.push(from(result))
      }

      if (R.is(Object)(data[key])) {
        _asyncMap(asyncFn, data[key], observables)
      }
    }
  }

  return data
}
/**
 * @param {(key: string,parent:any) => any} asyncFn
 * @param {Object} data
 * @returns {Promise<any>}
 */
export async function asyncMap(asyncFn, data) {
  const tempData = R.clone(data)
  /** @type {[Observable<any>]} */
  // @ts-ignore
  let observables = []
  // @ts-ignore
  _asyncMap(asyncFn, tempData, observables)

  await from(observables)
    .pipe(mergeAll())
    .toPromise()
  return tempData
}
/**
 * 将数组按个数分组
 *
 * @param {number} n
 * @param {Array} list
 * @returns {Array}
 */
function group(n, list) {
  return R.isEmpty(list)
    ? []
    : R.prepend(R.take(n, list), group(n, R.drop(n, list)))
}

export const groupsOf = R.curry(group)

/**
 * 将上游的数据以 {key,value} 的形式缓存,通过 key 区分数据
 *
 * 如果缓存中不存在,则运行生成函数获取数据,放入缓存,并传递给下游
 *
 * 如果存在,则直接从缓存中索引数据,传入下游
 *
 * @param {(value: any) => string} keySelector 可选函数,设定缓存的 key
 * @param {(value: any) => Promise<any>|Observable<any>} project 可选函数,value 生成函数,返回 Promise 或者流
 */
export function cache(keySelector = v => v, project = v => of(v)) {
  const values = new Map()
  /**
   * @param {Observable<any>} source
   */
  function cacheOperator(source) {
    return source.pipe(
      mergeMap(value => {
        const key = keySelector(value)
        if (!values.has(key)) {
          const result = from(project(value)).pipe(
            map(v => {
              values.set(key, v)
              return v
            }),
          )
          return result
        } else {
          return of(values.get(key))
        }
      }, 1),
    )
  }
  return cacheOperator
}
