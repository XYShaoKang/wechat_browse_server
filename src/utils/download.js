import fetch from 'node-fetch'

import { saveFile } from './file'
import { from } from 'rxjs'

/**
 * @param {string} url
 * @returns {Promise<{stream:NodeJS.ReadableStream; mimetype:string;}>}
 */
const getFile = (url, retryCount = 0) => {
  return fetch(url)
    .then(res => {
      return {
        stream: res.body,
        mimetype: res.headers.get('content-type') || '',
      }
    })
    .catch(err => {
      if (
        (err.message.includes('connect ETIMEDOUT') ||
          err.message.includes('read ECONNRESET')) &&
        retryCount < 3
      ) {
        retryCount++
        return new Promise(function(resolve) {
          setTimeout(() => {
            resolve()
          }, 1000)
        }).then(() => {
          return getFile(url, retryCount)
        })
      }
      throw new Error(err)
    })
}

/**
 * @param {string} url
 * @returns {Promise<{fileName:string; mimetype:string; size:number; url:string;}|null>} fileIndex id
 */
const downloadFile = async url => {
  if (!url) {
    return null
  }

  // 下载文件
  const fileStream = await getFile(url)

  // // 保存文件
  const { fileName, mimetype, size } = await saveFile(fileStream)
  return { fileName, mimetype, size, url }
}

/**
 * @param {string} url
 */
const download = url => from(downloadFile(url))

export { download, getFile }
