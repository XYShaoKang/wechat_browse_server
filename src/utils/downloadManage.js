import { Subject, from, of } from 'rxjs'
import {
  filter,
  pluck,
  first,
  map,
  scan,
  mergeMap,
  publishReplay,
  refCount,
  combineAll,
  flatMap,
} from 'rxjs/operators'

import { randomMd5, saveToFileIndex } from './file'
import { download } from './download'
import { cache } from './other'

const store = new Subject()
const thread = 20

export const download$ = (thread = 5) => {
  return mergeMap(
    /**
     * @param {{key:string,url:string}} arg
     */
    ({ key, url }) => from(download(url)).pipe(map(file => ({ key, file }))),
    thread,
  )
}
export const saveToFileIndex$ = () =>
  cache(
    ({ fileName }) => fileName,
    async file => {
      /**
       * @type {string}
       */
      const id = await saveToFileIndex(file)
      return id
    },
  )

const result$ = store.pipe(
  download$(thread),
  flatMap(({ key, file }) => {
    return of(file).pipe(
      saveToFileIndex$(),
      map(
        /**
         * @param {string} id
         */
        id => ({ key, id }),
      ),
    )
  }, 1),
  publishReplay(),
  refCount(),
)

from([
  result$.pipe(
    scan(
      /**
       * @param {number} acc
       * @param {{ key: string; id: string; }} _
       *
       * @returns {number}
       */
      // eslint-disable-next-line no-unused-vars
      (acc, _) => ++acc,
      0,
    ),
  ),
  store.pipe(scan(acc => ++acc, 0)),
])
  .pipe(combineAll())
  .subscribe(([curr, total]) => {
    if ((curr * 10) % total === 0) {
      // eslint-disable-next-line no-console
      console.log(`任务进度: ${(curr / total) * 100}% 总任务数: ${total}`)
    }
  })

/**
 * @param {string} url
 *
 * @returns {import('rxjs').Observable<string>} fileIndexId
 */
export const downloadManage = url => {
  const md5 = randomMd5()
  store.next({ key: md5, url })
  return result$.pipe(
    filter(({ key }) => key === md5),
    first(),
    pluck('id'),
  )
}
