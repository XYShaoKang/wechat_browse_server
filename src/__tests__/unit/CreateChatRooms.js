//#region mock

jest.mock('../../utils/download')
import * as downloadManageUtils from '../../utils/downloadManage'

let id = 0
const downloadManage = jest.fn(() => {
  return of(++id + '')
})

jest
  .spyOn(downloadManageUtils, 'downloadManage')
  .mockImplementation(downloadManage)

//#endregion

import { of } from 'rxjs'

import { Prisma, graphqlTestCall } from '../__utils'
import { CreateChatRooms } from './__types'

const variables = {
  chatRooms: [
    {
      username: 'aaaa',
      nickname: 'bbbb',
      displayName: 'cccc',
      owner: '',
      memberList: ['dddd'],
      modifyTime: new Date('2019-04-24T16:51:14.743Z'),
      avatar: {
        bigImg: 'http://a.com/b',
        thumbnailImg: 'http://b.com/a',
      },
    },
  ],
}

describe.only('Mutation', () => {
  /** @type {Prisma} */
  let prisma
  beforeEach(() => {
    prisma = new Prisma()
  })
  it('create ChatRoom', async () => {
    const result = await graphqlTestCall({
      query: CreateChatRooms,
      variables,
      context: {
        prisma,
        currentUser: { id: '0' },
        currentWeChat: { id: '0' },
      },
    })

    const id = result.data && result.data.CreateChatRooms.id

    expect(id).toBe('0')
    expect(prisma.updateWeChat.mock.calls.length).toBe(1)
    expect(downloadManage.mock.calls.length).toBe(2)
    expect(result).toMatchSnapshot()
    expect(prisma.db.chatRooms).toMatchSnapshot()
    expect(prisma.db.avatars).toMatchSnapshot()
  })
  it('create WeChatUser Fail No WeChat', async () => {
    const result = await graphqlTestCall({
      query: CreateChatRooms,
      variables,
      context: { prisma, currentUser: { id: '0' } },
    })
    const message = result.errors && result.errors[0].message

    expect(message).toBe('No WeChat')
    expect(result).toMatchSnapshot()
  })
})
