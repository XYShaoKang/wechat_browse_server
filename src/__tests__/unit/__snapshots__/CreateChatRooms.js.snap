// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation create ChatRoom 1`] = `
Object {
  "data": Object {
    "CreateChatRooms": Object {
      "id": "0",
    },
  },
}
`;

exports[`Mutation create ChatRoom 2`] = `
Array [
  Object {
    "avatar": Object {
      "id": "0",
    },
    "displayName": "cccc",
    "id": "0",
    "memberList": Array [
      Object {
        "id": "1",
      },
    ],
    "modifyTime": 2019-04-24T16:51:14.743Z,
    "nickname": "bbbb",
    "owner": "",
    "username": "aaaa",
  },
]
`;

exports[`Mutation create ChatRoom 3`] = `
Array [
  Object {
    "bigImg": Object {
      "id": "2",
    },
    "id": "0",
    "thumbnailImg": Object {
      "id": "1",
    },
  },
]
`;

exports[`Mutation create WeChatUser Fail No WeChat 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: No WeChat],
  ],
}
`;
